### PR TITLE
[DOCS-221] Best Pratices: Input Validation

### PR DESCRIPTION
**Pull request by mjang-cobalt**
Hi @sourcecodereviewer. 

Thanks so much for your pull request, https://github.com/cobalthq/cobalt-product-public-docs/pull/153

I've tried a rewrite of your content, and set it up in this PR. When I rewrote your content, I tried to:

- Keep your messaging
- Remove duplication
- Focus on [Active Voice](https://developers.google.com/style/voice)
- Include more specifics on your code samples

If I've made mistakes on any of this, please let me know.

You can review how this PR would look in our documentation in https://deploy-preview-154--cobalt-docs.netlify.app/bestpractices/akhan_input_validation/ (We've set up the PR to automatically build our content.)

(Note: I can't make you a "Reviewer" as we have limits on our GitHub subscription. However, you should still be able to review each line of the PR.)

### CODE REVIEW

File: content/en/BestPractices/akhan_input_validation.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+`product.limitPerUser`. However, it does not check if the quantity exceeds the:
+
+- Limit per user
+- Limit per order
+
+It even skips the check for deluxe users (`security.isDeluxe(req)`). However, it
+does *not* check if the quantity is positive.
+
+This small mistake is costly. Users can freely input negative values for
+`quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100
```
Comment (sourcecodereviewer) [inline content/en/BestPractices/akhan_input_validation.md:76]: changes the meaning slightly:

- attacker purchases `-100` units
- he needs to check out and use the `wallet` payment option 
- then (item cost * qty) gets deducted from his wallet - at this point, 100 times value of item is added to their wallet

File: content/en/BestPractices/akhan_input_validation.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+### Analysis
+
+The OWASP Juice Shop uses an [API
+Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)
+to call a
```
Comment (sourcecodereviewer) [inline content/en/BestPractices/akhan_input_validation.md:62]: The OWASP Juice Shop frontend initiates an API POST request to add items to the user's basket. The middleware function that is mounted on that API route handler (or API path) calls the `quantityCheck()` function.

File: content/en/BestPractices/akhan_input_validation.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
+function.
+
+It checks the quantity in the request payload with validation logic, using
+`product.limitPerUser`. However, it does not check if the quantity exceeds the:
```
Comment (sourcecodereviewer) [inline content/en/BestPractices/akhan_input_validation.md:67]: The `quantityCheck()` function checks if the quantity is valid and doesn't exceed the limit allowed per user or, the limit allowed per order. It bypasses the check if the user is a `deluxe` user.

**Review by sourcecodereviewer — CHANGES_REQUESTED** [review]
thanks, left comments on some areas that need to be slightly modified

**Comment by mjang-cobalt** [comment]
@sourcecodereviewer I love your comments! However, it might be Monday US time before I can respond / incorporate what you've suggested.

File: content/en/BestPractices/akhan_input_validation.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+}
+```
+
+### Analysis
+
+The OWASP Juice Shop uses an [API
+Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)
+to call a
+[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
+function.
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:64]:
```suggestion
The OWASP Juice Shop frontend initiates an [API
Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370) request to add items
to the user's basket. The middleware function that is mounted on that API route handler (or API path) calls the
[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
function.
```

File: content/en/BestPractices/akhan_input_validation.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+It checks the quantity in the request payload with validation logic, using
+`product.limitPerUser`. However, it does not check if the quantity exceeds the:
+
+- Limit per user
+- Limit per order
+
+It even skips the check for deluxe users (`security.isDeluxe(req)`). However, it
+does *not* check if the quantity is positive.
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:73]: @sourcecodereviewer How does this look? 

```suggestion
The `quantityCheck()` function validates the quantity in the request payload, as it checks:

- If the product is available (`const product = await QuantityModel.findOne({ where: { ProductId: id } })`)
- If the quantity exceeds the quantity available (`product.quantity >= quantity`)
- If the quantity exceeds the limit per user ((`product.limitPerUser && product.limitPerUser >= quantity)`) except for deluxe users (`security.isDeluxe(req)`)

However, it does **not** check if the quantity is a positive whole number.
```

File: content/en/BestPractices/akhan_input_validation.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+This small mistake is costly. Users can freely input negative values for
+`quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100
+times the value of the item to their wallet.
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:77]:
```suggestion
This small mistake is costly. Users can freely input negative values for
`quantity`. In that case:

- Assume an attacker adds `-100` units to their basket
  - The attacker then checks out of their shopping cart, with the `wallet` payment option
- The Juice Shop adds 100 times the value of the item to the wallet
```

File: content/en/BestPractices/akhan_input_validation.md (hunk 1 above)
Comment (mjang-cobalt) [reply content/en/BestPractices/akhan_input_validation.md:76]: I think I've corrected the text to match your meeting in the [next comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945700345)

File: content/en/BestPractices/akhan_input_validation.md (hunk 3 above)
Comment (mjang-cobalt) [reply content/en/BestPractices/akhan_input_validation.md:67]: See [next comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945688933) for response

File: content/en/BestPractices/akhan_input_validation.md (hunk 2 above)
Comment (mjang-cobalt) [reply content/en/BestPractices/akhan_input_validation.md:62]: @sourcecodereviewer I hope my suggested change[ in this comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945674503) matches what you're asking for

File: content/en/BestPractices/akhan_input_validation.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+### Impact
+
+Without such checks, this error could lead to substantial business losses.
+
+### Prevention
+
+In the following code sample, we include code that:
+
+- Checks if the quantity is greater than zero (`quantity > 0`)
```
Comment (sourcecodereviewer) [inline content/en/BestPractices/akhan_input_validation.md:87]: Thanks, @mjang, looks really good.
 
I wanted to edit the specific code in the bracket to `quantity <= 0` to match the code fix. This is my mistake.

File: content/en/BestPractices/akhan_input_validation.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+## Best Practices
+
```
Comment (sourcecodereviewer) [inline content/en/BestPractices/akhan_input_validation.md:116]: Two more points I think I missed:

- Always check if the user input matches the right format.
- Always ensure that the user input fits the given business context
  - A user should only enter a number in a quantity field, but the user should not be able to enter negative values there.
  - There should be sensible minimum and maximum values for a given quantity.

File: content/en/BestPractices/akhan_input_validation.md (hunk 7 above)
Comment (mjang-cobalt) [reply content/en/BestPractices/akhan_input_validation.md:87]: Oh duh right! I should have seen it too.

```suggestion
- Throws an error for invalid quantities (`quantity <= 0`)
```

File: content/en/BestPractices/akhan_input_validation.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+```ts
+// define the allow list regex for US Zip code
+const validZipPattern = /^\d{5}(-\d{4})?$/;
+// check if the user input matches our defined allow list regex
+if (!validZipPattern.test(userInput) {
+  // log the input validation failure
+  logger.warn("Invalid zip code provided.");
+  throw new Error("InvalidZipCode");
+}
+```
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:147]: Revise formatting per Markdown requirements

````suggestion
that you use input validation as early as possible in the data flow,
- Include an **allow list** of valid inputs. Attackers can often bypass the deny
  lists. For example, you could use the following to validate a US zip code:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
   		throw new Error("InvalidZipCode");
  }	
  ```

- Log all input validation failures. The information you get can help you detect
  intrusions, and identify attackers who try to send invalid inputs. For this
  code sample, you can add the following logging statement: `logger.warn("Invalid zip code provided.")`:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
    // log the input validation failure
    logger.warn("Invalid zip code provided.");
    throw new Error("InvalidZipCode");
  }
  ```
````

File: content/en/BestPractices/akhan_input_validation.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+## Best Practices
+
+- Assume that any input can be an attack
+- Validate all user input before processing it
+  - The [OWASP Input Validation Cheat Sheet
+    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
+that you use input validation as early as possible in the data flow,
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:121]:
```suggestion
    that you use input validation as early as possible in the data flow,
```

File: content/en/BestPractices/akhan_input_validation.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+}
+```
+
+## Best Practices
+
+- Assume that any input can be an attack
+- Validate all user input before processing it
+  - The [OWASP Input Validation Cheat Sheet
+    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
+    that you use input validation as early as possible in the data flow,
```
Comment (mjang-cobalt) [inline content/en/BestPractices/akhan_input_validation.md:121]: Fix punctuation for consistency

```suggestion
- Assume that any input can be an attack.
- Validate all user input before processing it.
  - The [OWASP Input Validation Cheat Sheet
    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
    that you use input validation as early as possible in the data flow.
```

File: content/en/BestPractices/akhan_input_validation.md (hunk 8 above)
Comment (mjang-cobalt) [reply content/en/BestPractices/akhan_input_validation.md:116]: I love examples!

I've rewritten what you've suggested slightly, and added it in https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r946719506

**Comment by mjang-cobalt** [comment]
Hi @sourcecodereviewer , I think I've addressed all of your requested changes.

IMO, this is ready to publish. Can you approve the PR?

**Review by sourcecodereviewer — APPROVED** [review]

**Comment by mjang-cobalt** [comment]
Previously approved before rebase. Merging on that basis.

**Pull request merged**